[MRG] Add Yeo-Johnson transform to PowerTransformer

NicolasHug opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#pull-request-checklist
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
Closes #10261 

#### What does this implement/fix? Explain your changes.

This PR implements the Yeo-Johnson transform as part of the PowerTransformer class.

PowerTransformer currently only support Box-Cox which only works for positive values, Yeo-Johnson works for the whole real line.

Original paper : [link](https://scholar.google.com/scholar?hl=en&as_sdt=0%2C44&q=https%3A%2F%2Fwww.jstor.org%2Fstable%2Fpdf%2F2673623.pdf&btnG=).

TODO:

* [X] Write transform
* [X] Fix lambda param estimation
* [X] Write inverse transform
* [X] Write docs
* [X] Write tests
* [X] Update examples

#### Any other comments?

The lambda parameter estimation is a bit tricky ~~and currently does not work.~~ (should be OK now, see below). Unlike for Box-Cox there's no scipy built-in that we can rely on. I'm having a hard time finding decent guidelines, tried to implement likelihood maximization with the brent optimizer (just like for Box-Cox) but run into overflow issues.

The transform code seems to work though:

![boxcox_yeojohnson](https://user-images.githubusercontent.com/1190450/42728453-bed2f218-8788-11e8-84ba-fb07360c0909.png)

which is a reproduction of

![blah](https://user-images.githubusercontent.com/1190450/42728468-1e2f222c-8789-11e8-976b-0356933d7b99.png)
From *Quantile regression via vector generalized additive models* by Thomas W. Yee.

Code for figure (hacky):
<details>

```python

import numpy as np
from sklearn.preprocessing import PowerTransformer
import matplotlib.pyplot as plt

yj = PowerTransformer(method='yeo-johnson', standardize=False)
bc = PowerTransformer(method='box-cox', standardize=False)

X = np.arange(-4, 4, .1).reshape(-1, 1)
fig, axes = plt.subplots(ncols=2)

for lmbda in (0, .5, 1, 1.5, 2):
    X_pos = X[X > 0].reshape(-1, 1)
    bc.fit(X_pos)
    bc.lambdas_ = [lmbda]
    X_trans = bc.transform(X_pos)
    axes[0].plot(X_pos, X_trans, label=r'$\lambda = {}$'.format(lmbda))
    axes[0].set_title('Box-Cox')

    yj.fit(X)
    yj.lambdas_ = [lmbda]
    X_trans = yj.transform(X)
    axes[1].plot(X, X_trans, label=r'$\lambda = {}$'.format(lmbda))
    axes[1].set_title('Yeo-Johnson')

for ax in axes:
    ax.set(xlim=[-4, 4], ylim=[-5, 5], aspect='equal')
    ax.legend()
    ax.grid()

plt.show()
```

</details>

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

NicolasHug committed:
WIP - First draft on Yeo-Johnson transform

NicolasHug committed:
Fixed lambda param optimization

The issue was from an error in the log likelihood function

NicolasHug commented:
Lambda param estimation should be fixed now, thanks @amueller.

Replication of [this example](http://scikit-learn.org/dev/auto_examples/preprocessing/plot_power_transformer.html) with Yeo-Johnson instead of Box-Cox:

![figure_1](https://user-images.githubusercontent.com/1190450/42729038-4043a62a-8799-11e8-9530-5a461f091f19.png)

NicolasHug committed:
Some first tests

Need to write inverse_transform to continue

NicolasHug committed:
Put helper method for yeo-johnson at the end

amueller commented on sklearn/preprocessing/data.py:
```diff
@@ -2491,10 +2492,15 @@ def fit(self, X, y=None):
+                # get rid of them to compute them.
+                _, lmbda = stats.boxcox(col[~np.isnan(col)], lmbda=None)
+                col_trans = boxcox(col, lmbda)
+            else:  # neo-johnson
```
yeo?

ogrisel replied:
Follow the right rabbit.

amueller replied:
this took me a while.

amueller commented on sklearn/preprocessing/data.py:
```diff
@@ -2557,6 +2566,53 @@ def inverse_transform(self, X):
+        # when x >= 0
+        if lmbda < 1e-19:
+            out[pos] = np.log(x[pos] + 1)
+        else:  #lmbda != 0
```
space after ``#``

amueller commented on sklearn/preprocessing/data.py:
```diff
@@ -2557,6 +2566,53 @@ def inverse_transform(self, X):
+            n = x.shape[0]
+
+            # Estimated mean and variance of the normal distribution
+            mu = psi.sum() / n
```
do we need ``from __future__ import division``?

NicolasHug replied:
it's here already

amueller replied:
thanks, was hard to see from the diff and I was lazy ;)

amueller commented on sklearn/preprocessing/tests/test_data.py:
```diff
@@ -2076,7 +2078,7 @@ def test_power_transformer_strictly_positive_exception():
 
     assert_raise_message(ValueError, not_positive_message,
-                         power_transform, X_with_negatives)
+                         power_transform, X_with_negatives, 'box-cox')
```
why is this needed? The default value shouldn't change, right? Or do we want to start a cycle to change the default to yeo-johnson?

NicolasHug replied:
I find it clearer and explicit?

I don't know if we'll change the default but it should still be fine as PowerTransform hasn't been released yet AFAIK

amueller replied:
good point. We should discuss before the release. I think yeo-johnson would make more sense.

ogrisel replied:
The fact that Yeo-Johnson accepts negative values while Box-Cox does not makes me feel like we should use it by default. From a usability point of view, it's nicer to our users.

NicolasHug replied:
I have the same feeling. Plus, it is designed to be a generalization of Box-Cox, even though that's not strictly the case.

amueller replied:
+1

NicolasHug replied:
shall I change the default then?

amueller replied:
think so.

NicolasHug replied:
Done

amueller reviewed (commented):
We think it's working now, right? So we need a test for the optimization, and then documentation and adding it to an example?

NicolasHug committed:
Added inverse transform + some tests

NicolasHug committed:
Added test for the optimization procedures

NicolasHug committed:
Created _box_cox_optimize method for better code symmetry

NicolasHug committed:
Opt for yeo-johnson not influenced by Nan

Also added related test

NicolasHug committed:
Added doc

NicolasHug committed:
Better test for nan in transform()

NicolasHug committed:
Updated more docs and example

ogrisel commented on sklearn/preprocessing/tests/test_data.py:
```diff
@@ -2122,3 +2150,59 @@ def test_power_transformer_lambda_zero():
+    pt = PowerTransformer(method=method, standardize=False)
+    pt.lambdas_ = [lmbda]
+    X_inv = pt.inverse_transform(X)
+    pt.lambdas_ = [9999]  # just to make sure
```
Why not:

```python
del pt.lambdas_
```

ogrisel replied:
Alternatively, create a new `pt` object from scratch to make the motivation of the test easier to read:

```python
ground_truth_transform = PowerTransformer(method=method, standardize=False)
ground_truth_transform.lambdas_ = [lmbda]
X_inv = pt.inverse_transform(X)

estimated_transform = PowerTransformer(method=method, standardize=False)
X_inv_trans = estimated_transform.fit_transform(X_inv) 
```

NicolasHug committed:
updated test

ogrisel commented on sklearn/preprocessing/tests/test_data.py:
```diff
@@ -2122,3 +2150,59 @@ def test_power_transformer_lambda_zero():
+    X_inv_trans = pt.fit_transform(X_inv)
+
+    assert_almost_equal(0, np.linalg.norm(X - X_inv_trans) / n_samples,
+                        decimal=2)
```
Please also add an assertion that checks that `X_inv_trans.mean(axis=0)` is close to `[0.]` and `X_inv_trans.std(axis=0)` is close to `[1.]`.

ogrisel commented on sklearn/preprocessing/tests/test_data.py:
```diff
@@ -2122,3 +2150,59 @@ def test_power_transformer_lambda_zero():
+
+    rng = np.random.RandomState(0)
+    n_samples = 1000
+    X = rng.normal(size=(n_samples, 1))
```
to make the test more explicit you can write: `X = rng.normal(loc=0., scale=1., size=(n_samples, 1))`

amueller commented:
If we want this to be the default then this is a blocker, right?

ogrisel commented on sklearn/preprocessing/tests/test_data.py:
```diff
@@ -2122,3 +2150,59 @@ def test_power_transformer_lambda_zero():
+    lmbda_no_nans = pt.lambdas_[0]
+
+    # concat nans at the end and check lambda stays the same
+    X = np.concatenate([X, np.full_like(X, np.nan)])
```
To make sure that the location of the NaNs does not impact the estimation:

```python
from sklearn.utils import shuffle
...

X = np.concatenate([X, np.full_like(X, np.nan)])
X = shuffle(X, random_state=0)
```

NicolasHug replied:
Done

ogrisel commented on sklearn/preprocessing/tests/test_data.py:
```diff
@@ -2122,3 +2150,59 @@ def test_power_transformer_lambda_zero():
+
+
+@pytest.mark.parametrize("method, lmbda", [('box-cox', .5),
+                                           ('yeo-johnson', .1)])
```
Could you add more values for `lmbda` for each method? E.g.:

```python
[
 ('box-cox', .1),
 ('box-cox', .5),
 ('yeo-johnson', .1),
 ('yeo-johnson', .5),
 ('yeo-johnson', 1.),
]

NicolasHug committed:
Modified tests according to reviews

NicolasHug committed:
Changed default method from cox-box to yeo-johnson

ogrisel commented on examples/preprocessing/plot_power_transformer.py:
```diff
@@ -1,26 +1,27 @@
-applied to six different probability distributions: Lognormal, Chi-squared,
-Weibull, Gaussian, Uniform, and Bimodal.
+The power transform is useful as a transformation in modeling problems where
+homoscedasticity and normality are desired. Below are examples of Box-Cox and
```
I don't understand what "modeling problems where homoscedasticity is desired".

ogrisel replied:
Actually, this statement seems to be correct:

http://article.sapub.org/10.5923.j.ajms.20180801.02.html

It might be interesting to try to come up with a good example to show this corrective effect in a (maybe synthetic) linear regression problem. However, this is probably outside of the scope of the current PR.

amueller commented:
tagged for 0.20 and added blocker label. I don't like that we keep adding stuff but if we want to make it default we should do it now.

glemaitre commented on sklearn/preprocessing/data.py:
```diff
@@ -2409,9 +2412,9 @@ class PowerTransformer(BaseEstimator, TransformerMixin):
-        The power transform method. Currently, 'box-cox' (Box-Cox transform)
-        is the only option available.
+    method : str, (default='yeo-johnson')
+        The power transform method. Available methods are 'box-cox' and
```
We can maybe have a bullet point list for each method referring to the reference section.

glemaitre commented on sklearn/preprocessing/data.py:
```diff
@@ -2490,12 +2494,18 @@ def fit(self, X, y=None):
         self.lambdas_ = []
         transformed = []
 
+        opt_fun = {'box-cox': self._box_cox_optimize,
```
I would have expect `func` instead of `fun` :)

glemaitre replied:
optim_function

glemaitre commented on sklearn/preprocessing/data.py:
```diff
@@ -2490,12 +2494,18 @@ def fit(self, X, y=None):
+        opt_fun = {'box-cox': self._box_cox_optimize,
+                   'yeo-johnson': self._yeo_johnson_optimize
+                   }[self.method]
+        trans_fun = {'box-cox': boxcox,
```
probably `transform_function` is not so long to be called

glemaitre commented on sklearn/preprocessing/data.py:
```diff
@@ -2547,16 +2571,112 @@ def inverse_transform(self, X):
+
+        return x_inv
+
+    def _yeo_johnson_transform(self, x, lmbda):
```
we cannot just define the forward transform and take `1 / _yeo_johnson_transform` for the inverse?

NicolasHug replied:
The inverse here means f^{-1}, not 1 / f

glemaitre commented on sklearn/preprocessing/data.py:
```diff
@@ -2547,16 +2571,112 @@ def inverse_transform(self, X):
+            """Return the negative log likelihood of the observed data x as a
+            function of lambda."""
+            psi = self._yeo_johnson_transform(x, lmbda)
+            n = x.shape[0]
```
n_samples instead

glemaitre commented on sklearn/preprocessing/data.py:
```diff
@@ -2547,16 +2571,112 @@ def inverse_transform(self, X):
+            """Return the negative log likelihood of the observed data x as a
+            function of lambda."""
+            psi = self._yeo_johnson_transform(x, lmbda)
+            n = x.shape[0]
```
Uhm missing x most probably

glemaitre replied:
Oh I see, can we pass x as an argument as well as in the optimize function?

NicolasHug replied:
yes this is a nested function so x is implicitely passed anyway

glemaitre commented on sklearn/preprocessing/data.py:
```diff
@@ -2547,16 +2571,112 @@ def inverse_transform(self, X):
+
+            # Estimated mean and variance of the normal distribution
+            mu = psi.sum() / n
+            sig_sq = np.power(psi - mu, 2).sum() / n
```
Stupid question: is `sig_sq` the variance? If this is the case, you might want to call it `var`

NicolasHug replied:
I was following the paper's notation. Should I use mean (or mean_) also then?

glemaitre reviewed (approved):
Couple of opened comments.
If I am not wrong we should have something in the common `estimator_checks` which force the input to be positive to work with box-cox. We probably want to change this behavior with we change the default.

NicolasHug commented:
I made a quick example to illustrate the use of Yeo-Johnson vs. Box-Cox + offset.

As Box-Cox only accepts positive data, one solution is to shift the data by a fixed offset value (typically min(data) + eps):

![test](https://user-images.githubusercontent.com/1190450/42765571-53d7cf88-88e6-11e8-9d31-0345e0386720.jpg)

One thing we see is that the "after offset and Box-Cox" isn't as symmetric as the eo-Johnson and most importantly the values are much higher.

Is it worth adding this as an example? TBH I wouldn't be able to mathematically or intuitively explain those results.

GaelVaroquaux commented:
+1 on comments by @glemaitre . Also, the tests are failing.

NicolasHug committed:
Addressed most comments from @glemaitre, fixed flake8

NicolasHug committed:
Removed box-cox specific checks in estimator_checks

The default is now Yeo-Johnson which can handle negative data

NicolasHug committed:
More explicit variable names for mean and variance

glemaitre commented on sklearn/preprocessing/data.py:
```diff
@@ -2547,16 +2578,112 @@ def inverse_transform(self, X):
+        out = np.zeros(shape=x.shape)
+        pos = (x >= 0)  # binary mask
+
+        # Note: we're comparing lmbda to 1e-19 instead of strict equality to 0.
```
Put this note in the docstring of the function.

You might want to do the same for the inverse transform as well.

glemaitre commented on sklearn/preprocessing/data.py:
```diff
@@ -2547,16 +2578,112 @@ def inverse_transform(self, X):
+        """Find and return optimal lambda parameter of the Yeo-Johnson
+        transform by MLE, for observed data x.
+
+        Like for coxbox, MLE is done via the brent optimizer.
```
box-cox

glemaitre commented on sklearn/preprocessing/data.py:
```diff
@@ -2547,16 +2578,112 @@ def inverse_transform(self, X):
+        def _neg_log_likelihood(lmbda):
+            """Return the negative log likelihood of the observed data x as a
+            function of lambda."""
+            transformed = self._yeo_johnson_transform(x, lmbda)
```
x_trans

glemaitre commented on sklearn/preprocessing/tests/test_data.py:
```diff
@@ -2003,13 +2004,29 @@ def test_quantile_transform_valid_axis():
 
+@pytest.mark.parametrize('standardize', [True, False])
+@pytest.mark.parametrize('X', [X_1col, X_2d])
+def test_power_transformer_inverse(standardize, X):
```
you can parameterize this test with the method as well and do

```python
X = np.abs(X) if method == 'box-cox'
```

It will reduce the code test.

glemaitre commented on sklearn/preprocessing/tests/test_data.py:
```diff
@@ -2085,11 +2103,22 @@ def test_power_transformer_strictly_positive_exception():
-                         power_transform, np.zeros(X_2d.shape))
+                         power_transform, np.zeros(X_2d.shape), 'box-cox')
+
+    # It should not raise any error for yeo-johnson
```
I would do an additional test by parametrizing X_train, X_test with different sign as you did.

glemaitre commented on sklearn/preprocessing/tests/test_data.py:
```diff
@@ -2122,3 +2151,68 @@ def test_power_transformer_lambda_zero():
+                                           ('yeo-johnson', 1.),
+                                           ])
+def test_optimization_power_transformer(method, lmbda):
+    """Test the optimization procedure
```
Probably you don't need the dosctest. You should instead my comments if you fill that your code is not self explanatory

glemaitre reviewed (changes requested):
Couple of comments

NicolasHug committed:
Addressed comments from glemaitre

NicolasHug committed:
Changed number of bins in plots to auto

NicolasHug committed:
Fixed Nan issues (ignored warnings)

Also removed other checks about positive data now that Yeo-Johnson is the
default

NicolasHug committed:
Fixed docstring example issue

NicolasHug committed:
Merge branch 'master' into yeojohnson

NicolasHug committed:
Updated whatsnew

glemaitre commented on doc/whats_new/v0.20.rst:
```diff
@@ -137,12 +137,14 @@ Preprocessing
   :issue:`10210` by :user:`Eric Chang <ericchang00>` and
-  :user:`Maniteja Nandana <maniteja123>`.
+  :user:`Maniteja Nandana <maniteja123>`, and :issue:`10261` by :user:`Nicolas
+  Hug <contact@nicolas-hug.com`.
```
you need to put your github account there

```python
:user:`Nicolas Hug <NicolasHug>`
```

glemaitre commented on doc/whats_new/v0.20.rst:
```diff
@@ -137,12 +137,14 @@ Preprocessing
+  homoscedasticity are desirable.
   :issue:`10210` by :user:`Eric Chang <ericchang00>` and
-  :user:`Maniteja Nandana <maniteja123>`.
+  :user:`Maniteja Nandana <maniteja123>`, and :issue:`10261` by :user:`Nicolas
```
issue refer to the number of the PR (I know that this is confusing) so it would be `11520`

glemaitre commented on examples/preprocessing/map_data_to_normal.py:
```diff
@@ -0,0 +1,128 @@
+"""
+
+# Author: Eric Chang <ericchang2017@u.northwestern.edu>
+# Author: Nicolas Hug <contact@nicolas-hug.com>
```
You don't need to add `#Author`, just indent under the previous name

glemaitre commented on sklearn/preprocessing/data.py:
```diff
@@ -2432,13 +2437,13 @@ class PowerTransformer(BaseEstimator, TransformerMixin):
     >>> print(pt.fit(data))
-    PowerTransformer(copy=True, method='box-cox', standardize=True)
+    PowerTransformer(copy=True, method='yeo-johnson', standardize=True)
     >>> print(pt.lambdas_)  # doctest: +ELLIPSIS
```
You don't need ellipsis

glemaitre commented on sklearn/preprocessing/data.py:
```diff
@@ -2432,13 +2437,13 @@ class PowerTransformer(BaseEstimator, TransformerMixin):
     >>> print(pt.lambdas_)  # doctest: +ELLIPSIS
-    [ 1.051... -2.345...]
+    [1.38668178e+00 5.93926346e-09]
     >>> print(pt.transform(data))  # doctest: +ELLIPSIS
```
no ellipsis then

glemaitre commented on sklearn/preprocessing/data.py:
```diff
@@ -2547,16 +2577,113 @@ def inverse_transform(self, X):
+        out = np.zeros(shape=x.shape, dtype=x.dtype)
+        pos = x >= 0  # binary mask
+
+        # Note: we're comparing lmbda to 1e-19 instead of strict equality to 0.
```
This comment in the docstring and in both transform

glemaitre commented on sklearn/preprocessing/tests/test_common.py:
```diff
@@ -36,25 +36,23 @@ def _get_valid_samples_by_column(X, col):
+     (RobustScaler(with_centering=False), robust_scale, True)]
 )
-def test_missing_value_handling(est, func, support_sparse, strictly_positive):
+def test_missing_value_handling(est, func, support_sparse):
```
You need to revert those changes and add the two cases for the 2 methods.

NicolasHug committed:
Addressed comments from glemaitre

NicolasHug committed:
Merge branch 'master' into yeojohnson

NicolasHug committed:
Fixed comment issue

NicolasHug committed:
Updated example

NicolasHug committed:
fixed minor typos

ogrisel reviewed (approved):
This LGTM, the example, and its explanations are very interesting ;)

ogrisel commented:
@glemaitre I think your comments were all addressed. Can you please update your review status?

NicolasHug committed:
Updated comment in whatsnew following ogrisel comments

ogrisel commented:
The example needs to be renamed to add a `plot_` prefix to follow the project naming convention.

NicolasHug committed:
Renamed plot example

ogrisel commented:
The matplotlib rendering of the examples is not suitable for sphinx gallery integration as such, see for instance the rendering by circle-ci:

https://28345-843222-gh.circle-artifacts.com/0/doc/auto_examples/preprocessing/plot_map_data_to_normal.html#sphx-glr-auto-examples-preprocessing-plot-map-data-to-normal-py

I would rather split that figure into 4 figures, one for each distribution. We could have 4 quarter panels for each:

- original distribution: top left
- quantile transformer: bottom left
- box-cox: top right
- yeo-johnson: bottom right

TomDLT commented on sklearn/preprocessing/data.py:
```diff
@@ -2547,16 +2577,120 @@ def inverse_transform(self, X):
+        # the computation of lambda is influenced by NaNs so we need to
+        # get rid of them
+        x = x[~np.isnan(x)]
+        # choosing backet -2, 2 like for boxcox
```
bracket ?

TomDLT commented on sklearn/preprocessing/tests/test_data.py:
```diff
@@ -2061,11 +2075,12 @@ def test_power_transformer_2d():
-def test_power_transformer_strictly_positive_exception():
+def test_power_transformer_boxcox_strictly_positive_exception():
+    # Exceptions should be raised for negative arrays and zero arrays when
+    # method is coxbox
```
box-cox

TomDLT reviewed (commented):
**Suggestion:** When calling `fit` and `transform` on the same data, the transform is done twice. We could implement a `fit_transform` method to do it only once.

Note that we would need the strict equivalence between `fit(X).transform(X)` and `fit_transform(X)`.

NicolasHug committed:
Should fix test for Python 2.7

NicolasHug committed:
Should fix example plot

NicolasHug committed:
Addressed comments from TomDLT

NicolasHug commented:
>Suggestion: When calling fit and transform on the same data, the transform is done twice. We could implement a fit_transform method to do it only once.

Good idea, I tried to implement it but I don't think it's possible:

Currently when calling `fit_transform`, the pipeline goes like this:
`X -> transform -> scaler_fit -> scaler_transform -> transform`

So `transform` is indeed called twice, but not on the same data. I don't think avoiding one of them is possible, but I may be missing something?

In any case I changed `fit` so that `transform` is only called when needed, that is when `standardize` is `True`.

ogrisel commented:
It's fine to call just transfom on the test.

The goal here is to call `fit_transform` on the training set instead of `fit` and then `transform` on the training set as there is a redundant computation that happens when subsequently calling `fit` and then `transform` on the same data.

TomDLT commented:
What I have in mind is:

```py
def fit(self, X):
    _ = self._fit_transform(X, transform=False)
    return self

def fit_transform(self, X):
    Xt = self._fit_transform(X, transform=True)
    return Xt

def transform(self, X):
    check_is_fitted(self, 'lambdas_')
    X = self._check_input(X, check_positive=True, check_shape=True,
                          copy=self.copy)
    Xt = self._transform_inplace(X)
    return Xt

def _fit_transform(self, X, transform=True):
    X = self._check_input(X, check_positive=True, check_method=True,
                          copy=self.copy or not transform)
    optim_function = ...
    self.lambdas_ = []
    for col in X.T:
        with np.errstate(invalid='ignore'):  # hide NaN warnings
             lmbda = optim_function(col)
             self.lambdas_.append(lmbda)
    self.lambdas_ = np.array(self.lambdas_)

    Xt = None
    if self.standardize or transform:
        Xt = self._transform_inplace(X)

    if self.standardize:
        self._scaler = StandardScaler()
        self._scaler.fit(Xt)
        if transform:
            Xt = self._scaler.transform(Xt)
    return Xt

def _transform_inplace(self, X):
    transform_function = ...
    for i, lmbda in enumerate(self.lambdas_):
        with np.errstate(invalid='ignore'):  # hide NaN warnings
            X[:, i] = transform_function(X[:, i], lmbda)
    return X
```

Therefore we reuse the transformed data computed for the standardization.

TomDLT commented:
In any case, we probably need a test to verify that:
- The input is not changed by `fit`, `transform` or `fit_transform` when `copy=True`.
- The input is not changed by `fit`, when `copy=False`.
- The input is changed by `transform` or `fit_transform`, when `copy=False`.

NicolasHug commented:
Isn't there a general test for all estimators for that?

ogrisel committed:
OPTIM implement fit_transform

NicolasHug committed:
Added test fit_transform() == fit().transform()

ogrisel commented:
I pushed a quick implementation for `fit_transform` but indeed it does not handle in place scaling properly, although it was also wrong when `standardize=True`: we also need to pass `self._scaler = StandardScaler(copy=self.copy)`.

ogrisel commented on sklearn/preprocessing/data.py:
```diff
@@ -2485,27 +2493,44 @@ def fit(self, X, y=None):
         -------
         self : object
         """
+        self._fit(X, y=y, force_compute_transform=False)
```
Maybe `force_transform` is enough.

NicolasHug committed:
Added tests for the copy parameter

NicolasHug committed:
Fixed flake8 issues in example plot

TomDLT commented on sklearn/preprocessing/data.py:
```diff
@@ -2512,25 +2515,22 @@ def _fit(self, X, y=None, force_compute_transform=False):
-            self._scaler = StandardScaler()
-            if force_compute_transform:
-                transformed = self._scaler.fit_transform(transformed)
+            self._scaler = StandardScaler(copy=self.copy)
```
actually you should be able to use `copy=False` here, since a copy has already been done just before.

TomDLT reviewed (approved):
Thanks for the added tests.
We might want to add them as common tests at some point, but it might be for another pull-request.

NicolasHug committed:
set copy to False for the scaler

NicolasHug committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into yeojohnson

glemaitre commented on sklearn/preprocessing/data.py:
```diff
@@ -2485,27 +2493,44 @@ def fit(self, X, y=None):
+                                  'yeo-johnson': self._yeo_johnson_transform
+                                  }[self.method]
+            for i, lmbda in enumerate(self.lambdas_):
+                with np.errstate(invalid='ignore'):  # hide NaN warnings
```
I would move this inside the transform_function

glemaitre commented on sklearn/preprocessing/data.py:
```diff
@@ -2514,12 +2539,21 @@ def transform(self, X):
+                              }[self.method]
         for i, lmbda in enumerate(self.lambdas_):
-            X[:, i] = boxcox(X[:, i], lmbda)
+            with np.errstate(invalid='ignore'):  # hide NaN warnings
```
I would move this inside the transform function

glemaitre commented on sklearn/preprocessing/data.py:
```diff
@@ -2536,27 +2570,147 @@ def inverse_transform(self, X):
-            else:
-                x_inv = (x * lmbda + 1) ** (1 / lmbda)
-            X[:, i] = x_inv
+            with np.errstate(invalid='ignore'):  # hide NaN warnings
```
Inside the inverse function?

glemaitre commented on sklearn/preprocessing/data.py:
```diff
@@ -2536,27 +2570,147 @@ def inverse_transform(self, X):
+        """Return inverse-transformed input x following Yeo-Johnson inverse
+        transform with parameter lambda.
+
+        Note
```
Notes

glemaitre commented on sklearn/preprocessing/data.py:
```diff
@@ -2536,27 +2570,147 @@ def inverse_transform(self, X):
+        """Return transformed input x following Yeo-Johnson transform with
+        parameter lambda.
+
+        Note
```
Notes

glemaitre commented on sklearn/preprocessing/data.py:
```diff
@@ -2566,7 +2720,8 @@ def _check_input(self, X, check_positive=False, check_shape=False,
         check_positive : bool
-            If True, check that all data is positive and non-zero.
+            If True, check that all data is positive and non-zero (only if
+            self.method is box-cox).
```
only if ``self.method=='box-cox'``

glemaitre commented on sklearn/preprocessing/data.py:
```diff
@@ -2485,27 +2493,44 @@ def fit(self, X, y=None):
-            self.lambdas_.append(lmbda)
-            transformed.append(col_trans)
-
+            with np.errstate(invalid='ignore'):  # hide NaN warnings
```
In the optim_func

glemaitre reviewed (changes requested):
Couple of changes

NicolasHug committed:
Addressed comments from glemaitre

glemaitre commented:
I am waiting to check the example in the documentation

NicolasHug committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into yeojohnson

NicolasHug commented:
@glemaitre @ogrisel, I think [the plot](https://28921-843222-gh.circle-artifacts.com/0/doc/auto_examples/preprocessing/plot_map_data_to_normal.html#phx-glr-auto-examples-preprocessing-plot-map-data-to-normal-py) look pretty OK now.

jnothman commented:
I'm finding the plots in `plot_map_data_to_normal` relatively hard to navigate intuitively. It's not a blocker, but I think it needs to look more tabular: at the moment it takes some effort to see that each row is a different transformation; a label on the left of the row would be more helpful.

Also, having the transformations go from left to right and the datasets from top to bottom doesn't look like it would be infeasible, and would be more familiar from plot_cluster_comparison etc.

jnothman reviewed (commented):
Can I clarify why `plot_all_scaling` still only shows box-cox?

NicolasHug commented:
>Also, having the transformations go from left to right and the datasets from top to bottom doesn't look like it would be infeasible, and would be more familiar from plot_cluster_comparison etc.

Personally I find it easier to compare the transformations when they're stacked on each other, especially since the axes limits are uniform across the plots.

I don't have anything against having the transformation names on the left. It would also make sense to me to have one dataset per column (limiting the plot to 4 rows instead of 8), but that would make the plot wider which can be annoying on mobile.

Thanks for mentioning `plot_all_scaling`, I missed that one.

NicolasHug committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into yeojohnson

NicolasHug commented:
Looks like 14e7c328df9968ca403e47168314ae28390c07d5 broke `plot_all_scaling` on master:

```
Traceback (most recent call last):
  File "examples/preprocessing/plot_all_scaling.py", line 71, in <module>
    dataset = fetch_california_housing()
  File "/home/nico/dev/sklearn/sklearn/datasets/california_housing.py", line 128, in fetch_california_housing
    cal_housing = joblib.load(filepath)
  File "/home/nico/dev/sklearn/sklearn/externals/joblib/numpy_pickle.py", line 578, in load
    obj = _unpickle(fobj, filename, mmap_mode)
  File "/home/nico/dev/sklearn/sklearn/externals/joblib/numpy_pickle.py", line 508, in _unpickle
    obj = unpickler.load()
  File "/usr/lib64/python3.6/pickle.py", line 1050, in load
    dispatch[key[0]](self)
  File "/usr/lib64/python3.6/pickle.py", line 1338, in load_global
    klass = self.find_class(module, name)
  File "/usr/lib64/python3.6/pickle.py", line 1388, in find_class
    __import__(module, level=0)
ModuleNotFoundError: No module named 'sklearn.externals._joblib.numpy_pickle'

```

amueller commented:
@NicolasHug it's now "fixed" but you need to remove your ``scikit_learn_data`` folder in your home folder.

NicolasHug committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into yeojohnson

NicolasHug committed:
Updated plot_all_scaling.py  example

ogrisel commented:
The matplotlib rendering of the 2 examples is good now:

https://29575-843222-gh.circle-artifacts.com/0/doc/auto_examples/index.html#preprocessing

Merging. Thanks @NicolasHug for this nice contribution!

amueller commented:
yay!

ogrisel commented:
I agree that with the @jnothman (https://github.com/scikit-learn/scikit-learn/pull/11520#issuecomment-406407630) that using a layout similar to the cluster comparison plot would improve the readability even further but I don't want to delay the release for this.

GaelVaroquaux commented:
Yey!!
